Maven plugin does not work with Java 9

Hi,

in Vert.x we have integrated the rocker template engine. We are currently testing our builds with Java 9 and we found an issue when running the rocker-maven-plugin with Java 9:

```
[ERROR] Failed to execute goal com.fizzed:rocker-maven-plugin:0.22.0:generate (generate-rocker-templates) on project vertx-web-templ-rocker: Execution generate-rocker-templates of goal com.fizzed:rocker-maven-plugin:0.22.0:generate failed: begin 0, end 3, length 1 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.fizzed:rocker-maven-plugin:0.22.0:generate (generate-rocker-templates) on project vertx-web-templ-rocker: Execution generate-rocker-templates of goal com.fizzed:rocker-maven-plugin:0.22.0:generate failed: begin 0, end 3, length 1
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution generate-rocker-templates of goal com.fizzed:rocker-maven-plugin:0.22.0:generate failed: begin 0, end 3, length 1
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end 3, length 1
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3116)
	at java.base/java.lang.String.substring(String.java:1885)
	at com.fizzed.rocker.maven.GenerateMojo.execute(GenerateMojo.java:117)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
```

--- a/rocker-compiler/src/main/java/com/fizzed/rocker/compiler/JavaGenerator.java
+++ b/rocker-compiler/src/main/java/com/fizzed/rocker/compiler/JavaGenerator.java
@@ -28,7 +28,6 @@ import java.nio.file.Path;
 import java.util.*;
 
 import static com.fizzed.rocker.compiler.RockerUtil.*;
-import javax.annotation.Generated;
 
 public class JavaGenerator {
     static private final Logger log = LoggerFactory.getLogger(JavaGenerator.class);
@@ -169,7 +168,6 @@ public class JavaGenerator {
        
         // imports regardless of template
         w.append(CRLF);
-        w.append("import ").append(Generated.class.getName()).append(";").append(CRLF);
         w.append("import ").append(java.io.IOException.class.getName()).append(";").append(CRLF);
         w.append("import ").append(com.fizzed.rocker.ForIterator.class.getName()).append(";").append(CRLF);
         w.append("import ").append(com.fizzed.rocker.RenderingException.class.getName()).append(";").append(CRLF);
@@ -202,7 +200,7 @@ public class JavaGenerator {
         // MODEL CLASS
         
         // annotations (https://github.com/fizzed/rocker/issues/59)
-        tab(w, indent).append("@Generated(\"com.fizzed.rocker.compiler.JavaGenerator\") @SuppressWarnings(\"unused\")")
+        tab(w, indent).append("@SuppressWarnings(\"unused\")")
             .append(CRLF);
         
         // class definition

--- a/rocker-compiler/src/main/java/com/fizzed/rocker/compiler/TemplateCompiler.java
+++ b/rocker-compiler/src/main/java/com/fizzed/rocker/compiler/TemplateCompiler.java
@@ -107,8 +107,17 @@ public class TemplateCompiler {
 
         // under maven or other build tools, java.class.path is wrong
         // build our own from current context
+        // todo: there doesn't seem to be any test checking this assumption is correct.
         StringBuilder classpath = new StringBuilder();
-        URL[] classpathUrls = ((URLClassLoader)(Thread.currentThread().getContextClassLoader())).getURLs();
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+
+        // starting in java 9, the class loader isn't necessarily an instance of URLClassLoader
+        URL[] classpathUrls = new URL[0];
+        if (contextClassLoader instanceof URLClassLoader) {
+            classpathUrls = ((URLClassLoader) contextClassLoader).getURLs();
+        } else {
+            classpath.append(System.getProperty("java.class.path"));
+        }
         for (URL url : classpathUrls) {
             if (classpath.length() > 0) {
                 classpath.append(File.pathSeparator);

--- a/rocker-compiler/src/main/java/com/fizzed/rocker/model/JavaVersion.java
+++ b/rocker-compiler/src/main/java/com/fizzed/rocker/model/JavaVersion.java
@@ -23,8 +23,10 @@ public enum JavaVersion {
     
     v1_6 (6, "1.6"),
     v1_7 (7, "1.7"),
-    v1_8 (8, "1.8");
-    
+    v1_8 (8, "1.8"),
+    V9   (9, "9"),
+    ;
+
     private final int version;
     private final String label;
     

--- a/rocker-maven-plugin/src/main/java/com/fizzed/rocker/maven/GenerateMojo.java
+++ b/rocker-maven-plugin/src/main/java/com/fizzed/rocker/maven/GenerateMojo.java
@@ -114,7 +114,18 @@ public class GenerateMojo extends AbstractMojo {
         
         if (javaVersion == null || javaVersion.length() == 0) {
             // set to current jdk version
-            javaVersion = System.getProperty("java.version").substring(0, 3);
+            String version = System.getProperty("java.version");
+            if (version.startsWith("1.")) {
+                javaVersion = version.substring(0, 3);
+            } else {
+                try {
+                    // javaVersion = String.valueOf(Runtime.version().major());
+                    Object runtimeVersion = Runtime.class.getMethod("version").invoke(null);
+                    javaVersion = runtimeVersion.getClass().getMethod("major").invoke(runtimeVersion).toString();
+                } catch (Exception e) {
+                    throw new MojoExecutionException("Could not infer java version, set the rocker.javaVersion property manually to fix this problem", e);
+                }
+            }
             getLog().info("Property rocker.javaVersion not set. Using your JDK version " + this.javaVersion);
         } else {
             getLog().info("Targeting java version " + this.javaVersion);

